File: Problems/39. Combination Sum.py
"""
Problem Link: https://leetcode.com/problems/combination-sum/
"""


class Solution:

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def combinationSum(candidates, target, index, current, result):
            if (target == 0):
                result.append(list(current))
                return
            if (target < 0):
                return
            if (index >= len(candidates)):
                return
            current.append(candidates[index])
            combinationSum(candidates, target -
                           candidates[index], index, current, result)
            current.pop()
            combinationSum(candidates, target, index + 1, current, result)
        result = []
        combinationSum(candidates, target, 0, [], result)
        return result

File: Problems/39. Combination Sum.java
/*
Problem Link: https://leetcode.com/problems/combination-sum/
*/

import java.util.*;

class Solution {
    private void combinationSum(
            int[] candidates, int target, int index, List<Integer> current, List<List<Integer>> result) {
        if (target == 0) {
            result.add(new ArrayList<>(current));
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = index; i < candidates.length; i++) {
            current.add(candidates[i]);
            combinationSum(candidates, target - candidates[i], i, current, result);
            current.remove(current.size() - 1);
        }

    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        combinationSum(candidates, target, 0, new ArrayList<>(), result);
        return result;
    }
}